### ISSUE TITLE
Manticore detects a false positive integer overflow in a very simple contract

### ISSUE THREAD

**Issue by ggrieco-tob** [trailofbits/manticore#714]
### OS / Environment

Distributor ID:	ArchLinux
Description:	ArchLinux
Release:	rolling-release

### Manticore version

Last revision (465c926a2e39d5daf1405711dac61b6ddd0a2527)

### Python version

Python 2.7.13

### Summary of the problem

Manticore detects a false positive integer overflow in a very simple contract.

### Step to reproduce the behavior

This is a small solidity contract to reproduce this issue:

```
contract C {
    function() {
        if (msg.data[0] == 1) {
            return;
        }else{
            return;
        }
    }
}
```

You can trigger this issue using the manticore cli:

```
$ manticore error.sol                                                                
```

### Expected behavior

It shouldn't detect an integer overflow.

### Actual behavior

```
2018-01-16 08:42:29,596: [13352] m.main:INFO: Beginning analysis
2018-01-16 08:42:29,633: [13352] m.ethereum:INFO: Starting symbolic transaction: 1
2018-01-16 08:42:29,969: [13352] m.ethereum:WARNING: Integer overflow at MUL instruction
2018-01-16 08:42:30,099: [13352] m.ethereum:INFO: Generated testcase No. 0 - REVERT
2018-01-16 08:42:33,105: [13352] m.ethereum:INFO: Generated testcase No. 1 - REVERT
2018-01-16 08:42:35,147: [13352] m.ethereum:INFO: Finished symbolic transaction: 1 | Code Coverage: 98% | Terminated States: 2 | Alive States: 2
2018-01-16 08:42:35,153: [13352] m.ethereum:INFO: Starting symbolic transaction: 2
2018-01-16 08:42:35,665: [13352] m.ethereum:INFO: Generated testcase No. 2 - REVERT
2018-01-16 08:42:39,842: [13352] m.ethereum:INFO: Generated testcase No. 3 - REVERT
2018-01-16 08:42:43,122: [13352] m.ethereum:INFO: Generated testcase No. 4 - REVERT
2018-01-16 08:42:47,779: [13352] m.ethereum:WARNING: Integer overflow at MUL instruction
2018-01-16 08:42:48,104: [13352] m.ethereum:INFO: Generated testcase No. 5 - REVERT
2018-01-16 08:42:51,469: [13352] m.ethereum:WARNING: Integer overflow at MUL instruction
2018-01-16 08:42:51,703: [13352] m.ethereum:INFO: Finished symbolic transaction: 2 | Code Coverage: 98% | Terminated States: 6 | Alive States: 4
2018-01-16 08:42:51,718: [13352] m.ethereum:INFO: Generated testcase No. 6 - STOP
2018-01-16 08:42:54,907: [13352] m.ethereum:INFO: Generated testcase No. 7 - STOP
2018-01-16 08:42:57,973: [13352] m.ethereum:INFO: Generated testcase No. 8 - STOP
2018-01-16 08:43:01,252: [13352] m.ethereum:INFO: Generated testcase No. 9 - STOP
2018-01-16 08:43:04,389: [13352] m.ethereum:INFO: Results in mcore_ttp4py
```

**Comment by mossberg** [comment]
thanks again for the issue @ggrieco-tob  :)

The logic is incorrect in the int overflow multiplication check. this is because we reused the same detection logic from add overflow :/
https://github.com/trailofbits/manticore/blob/master/manticore/ethereum.py#L61

`if state.can_be_true(result < arguments[0]) or state.can_be_true(result < arguments[1]):`

for a* b = c the logic needs to check if (c < a AND c < b) ( consider `0x?? * 0x100`; if 0x?? == 0, then the result (0) can be less than 0x100, but this would trigger the detector)

additionally, for multiplication, the checks are incorrect as they used a signed less than comparison, rather than unsigned.

**Comment by ggrieco-tob** [comment]
This is a small solidity contract that produces some arithmetic instructions (I don't know exactly why) and can cause false positives: 

```
contract C {

  function f(address[] x) returns (address[]) {
    return x;
  }
}
```

This contact cannot be run in manticore (master). It needs to use a patched version using the dev-evm-dynamicarguments but resolving some conflicts. I will re-open this tickets as soon as this example can be executed in the master branch.

**Comment by mossberg** [comment]
Some initial notes on this. so we get some warnings like this

```
2018-03-15 16:08:55,422: [121226] m.ethereum:WARNING: Integer overflow at MUL instruction
2018-03-15 16:08:55,592: [121226] m.ethereum:WARNING: Integer overflow at ADD instruction 6f
2018-03-15 16:08:55,792: [121226] m.ethereum:WARNING: Integer overflow at ADD instruction 75
2018-03-15 16:08:55,978: [121226] m.ethereum:WARNING: Integer overflow at MUL instruction

```

these warnings occur in code that is emitted by solc related to the `address[]` parameter. solc emits code that parses the data according to the ABI, computes the size of the passed array, and then does a calldatacopy to copy it into memory. the overflow warnings appear in the code for computing the size of the array.

here's some output from rattle thats a bit easier to read. from an initial look, these are not false positives, but i haven't spent enough time to totally confirm

```
  <SSA::BasicBlock ofs:0x51 insns:[
    %2 = CALLDATALOAD(#0x4),
    %3 = ADD(#0x20, #0x4),
    %4 = ADD(#0x4, %2), // %4 is head element; offset
    %5 = CALLDATALOAD(%4), // %5 is nelements
    %6 = ADD(#0x20, %4), // can overflow.
    %7 = MUL(#0x20, %5),
    %8 = ADD(#0x20, %7),
    %9 = MLOAD(#0x40),// 9 is the free ptr
    %a = ADD(%9, %8), //a == 9
    MSTORE(#0x40, %a),
    MSTORE(%9, %5),
    %b = ADD(#0x20, %9),
    %c = MUL(#0x20, %5), // nelements * 32 = total bytes for args
    CALLDATACOPY(%b, %6, %c),

```

**Comment by defunctio** [comment]
Verified this is resolved in master; 1f74f0fa84a1aa3911c68df1b9cf20800dd7c778.
Closing
